packages/material-ui/src/styles/colorManipulator.js

```diff
@@ -31,7 +31,15 @@ function clamp(value, min = 0, max = 1) {
  * @param {string} color - Hex color, i.e. #nnn or #nnnnnn
  * @returns {string} A CSS rgb color string
  */
-export function convertHexToRGB(color) {
+export function hexToRgb(color) {
+  if (process.env.NODE_ENV !== 'production') {
+    if (!color) {
+      throw new Error(
+        `Material-UI: hexToRgb(color) is called with an invalid argument. (${color})`,
```

Comment:
fast feedback for wrong usages

Reply:
This adds different behavior in production. dev and prod shouldn't behave differently. 

Either throw in dev too or use `warning`

Reply:
It should be fine. It will throw in development and production. The only difference is the error message.

Reply:
Never change control flow between environments. It is not equivalent.

Reply:
And even if: Why change the pattern? We always used `warning`. Why switch to `throw` here?

Reply:
The objective is to make it obvious that something is wrong. We throw a nice error before the code throw or we warn with a nice message. There are different examples of it in the codebase https://github.com/mui-org/material-ui/blob/534c40cee964cf7bfff6a013b3671d0ea293123b/packages/material-ui/src/utils/helpers.js#L9. The problem with throw + warning is the people see the throw first. With CRA, Next.js or Gatsby they will see the error message in the body of the page, they warning is only in the console.

Reply:
Now, I think that the real concern is: Should it throw in production or not? How critical is the color manipulation for an applications?

Reply:
It throws anywhere. For throw there is no distinction between environments. Not in react or anywhere I know because most devs agree that control flow should not change between environments.

Reply:
> Should it throw in production or not?

If not then it's a warning.

Reply:
I guess the concern is bundle size here. We could transform the messages into error codes with a link to material-ui.com that displays the full error message. `react` is doing this as well with  https://reactjs.org/docs/error-decoder.html/

Reply:
I think that we should answer this question first: What should happen in production when the params are invalid? We have 3 options:
1. We implement a defensive logic to concern the method in a no-op.
2. We do nothing, the logic breaks with an obscure error message.
3. We handle the exception, provide a nice message, with minification or not. React minification logic was implemented by an intern, it's probably overkilling at our scale.

Once we have answered this question, we can consider the development env. I would propose respectively:
1. We raise a warning
2. / 3 We change the control flow, throw a nice warning

This problem remembers me what happen with a few of our components when none of the required props are provided. [It fails with an obscure error message](https://github.com/oliviertassinari/material-ui-generate-css-file/blob/master/src/static-styles.js) and a warning is throw in the console:
> Invariant Violation: React.cloneElement(...): The argument must be a React element, but you passed undefined.

Reply:
For new react devs I agree. But a fundamental part of react debugging is the console. We can't solve any issues devs have here because react is dictating how it should work. The error overlay you cited even includes the solution to the issue: 
> Open your browser’s developer console to further inspect this error.

> Once we have answered this question, we can consider the development env

You always start with a premise which I don't agree with and react does not agree with and I think most other people won't agree with: Thrown errors are different in prod and dev. Why should a program behave differently in production or dev (and please don't move the goalpost to warnings they're bad enough)?

This is my 2nd attempt at focusing your attention on this point. If you insist on not discussing this I'm not willing to have a discussion about the next point either.

Reply:
> Why should a program behave differently in production or dev

I'm gonna state it a 2nd time. The behavior is identical. Only the error message change, like it change with React invariant.

Reply:
The react comparison is invalid. They don't use their invariants for simple types.

> The behavior is identical: 

Throwing is not always equivalent. This only applies for this scenario but is generally not true for e.g. side-effects.

> it throws

Also not true:
```js
hexToRgb(1);
hexToRgb({ substr: () => {}})
```

Reply:
Ok, I'm changing the approach.